exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/ConvertCountToDirectScan.java

```diff
@@ -85,109 +91,231 @@ protected ConvertCountToDirectScan(RelOptRuleOperand rule, String id) {
+          cnt = totalRecordCount;
+
+        } else if (aggCall.getArgList().size() == 1) {
+          // count(columnName) ==> Agg ( Scan )) ==> columnValueCount
+          int index = aggCall.getArgList().get(0);
+
+          if (project != null) {
+            // project in the middle of Agg and Scan : Only when input of AggCall is a RexInputRef in Project, we find the index of Scan's field.
+            // For instance,
+            // Agg - count($0)
+            //  \
+            //  Proj - Exp={$1}
+            //    \
+            //   Scan (col1, col2).
+            // return count of "col2" in Scan's metadata, if found.
+            if (!(project.getProjects().get(index) instanceof RexInputRef)) {
+              return false; // do not apply for all other cases.
+            }
+
+            index = ((RexInputRef) project.getProjects().get(index)).getIndex();
+          }
+
+          String columnName = scan.getRowType().getFieldNames().get(index).toLowerCase();
+
+          // for implicit column count will the same as total record count
+          if (implicitColumnsNames.contains(columnName)) {
+            cnt = totalRecordCount;
+          } else {
+            SchemaPath simplePath = SchemaPath.getSimplePath(columnName);
+            columns.add(simplePath);
+
+            cnt = oldGrpScan.getColumnValueCount(simplePath);
+            if (cnt == GroupScan.NO_COLUMN_STATS) {
+              // if column stats is not available don't apply this rule
+              return false;
+            }
+          }
+        } else {
+          return false;
+        }
+        counts.add(cnt);
+      }
+      return true;
+    }
 
-  private GroupScan getCountDirectScan(long cnt) {
-    CountQueryResult res = new CountQueryResult(cnt);
+    /**
+     * Checks if aggregate call contains star or non-null expression:
+     * <pre>
+     * count(*)  == >  empty arg  ==>  rowCount
+     * count(Not-null-input) ==> rowCount
+     * </pre>
+     *
+     * @param aggregateCall aggregate call
+     * @param aggregate aggregate relation expression
+     * @return true of aggregate call contains star or non-null expression
+     */
+    private boolean containsStarOrNotNullInput(AggregateCall aggregateCall, DrillAggregateRel aggregate) {
+      return aggregateCall.getArgList().isEmpty() ||
+          (aggregateCall.getArgList().size() == 1 &&
+              !aggregate.getInput().getRowType().getFieldList().get(aggregateCall.getArgList().get(0)).getType().isNullable());
+    }
 
-    PojoRecordReader<CountQueryResult> reader = new PojoRecordReader<CountQueryResult>(CountQueryResult.class,
-        Collections.singleton(res).iterator());
+    /**
+     * Checks if stores list of columns contains file system partition columns.
+     *
+     * @return true if contains partition columns, false otherwise
+     */
+    private boolean containsPartitionColumns() {
+      final ColumnExplorer columnExplorer = new ColumnExplorer(settings.getOptions(), columns);
```

Comment:
Can ColumnExplorer provide an API to tell if one particular column is the partition column? Seems to me that with such API, we could exit from the for loop in line 238, in stead of first building column list, then call this method to check if the column lists contain a file system partition column.